ambry-tools/src/main/java/com.github.ambry.tools/perf/rest/PerfRSC.java

```diff
@@ -0,0 +1,139 @@
+
+  private volatile AsyncWritableChannel writeChannel = null;
+  private volatile FutureResult<Long> futureResult = new FutureResult<Long>();
+  private volatile Callback<Long> readIntoCallback = null;
+
+  /**
+   * Create an instance of PerfRSC.
+   * @param bytes the bytes that form a chunk that will be repeatedly returned until {@code streamSize} is reached.
+   * @param streamSize the desired total size of the stream.
+   */
+  public PerfRSC(byte[] bytes, long streamSize) {
+    if (streamSize < 0 || (streamSize > 0 && bytes.length < 1)) {
+      throw new IllegalArgumentException("Invalid argument(s)");
+    }
+    this.streamSize = streamSize;
+    if (bytes != null) {
+      this.bytes = ByteBuffer.wrap(bytes);
+    } else {
+      this.bytes = null;
+    }
+  }
+
+  @Override
+  public long getSize() {
+    return streamSize;
+  }
+
+  @Override
+  public Future<Long> readInto(AsyncWritableChannel asyncWritableChannel, Callback<Long> callback) {
+    if (readIntoInvoked.compareAndSet(false, true)) {
+      writeChannel = asyncWritableChannel;
+      readIntoCallback = callback;
+      writeToChannel();
+    } else {
+      throw new IllegalStateException("ReadInto should not be called more than once");
+    }
+    return futureResult;
+  }
+
+  @Override
+  public void setDigestAlgorithm(String digestAlgorithm)
+      throws NoSuchAlgorithmException {
+    // no op
+  }
+
+  @Override
+  public byte[] getDigest() {
+    return null;
+  }
+
+  @Override
+  public boolean isOpen() {
+    return isOpen.get();
+  }
+
+  @Override
+  public void close() {
+    isOpen.set(false);
+  }
+
+  /**
+   * Writes continuously to the provided {@link AsyncWritableChannel} until {@link #streamSize} is reached.
+   */
+  private void writeToChannel() {
+    if (channelWriteException == null && streamed < streamSize) {
+      if (!isOpen()) {
+        invokeCallback(streamed, new ClosedChannelException());
+      } else {
+        int bytesLeft = streamSize - streamed > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) (streamSize - streamed);
+        bytes.rewind();
+        bytes.limit(Math.min(bytes.capacity(), bytesLeft));
+        writeChannel.write(bytes, new Callback<Long>() {
+          @Override
+          public void onCompletion(Long result, Exception callbackException) {
+            streamed += result;
+            channelWriteException = callbackException;
+            writeToChannel();
+          }
+        });
+      }
+    } else {
+      invokeCallback(streamed, channelWriteException);
+    }
+  }
+
+  /**
+   * Invokes the callback and updates the future once this is called. This function ensures that the callback is invoked
+   * just once.
+   * @param bytesRead the number of bytes read.
+   * @param exception the {@link Exception}, if any, to pass to the callback.
+   */
+  private void invokeCallback(long bytesRead, Exception exception) {
+    if (callbackInvoked.compareAndSet(false, true)) {
```

Comment:
Yes. two `close()` may be called at the same time or a `close()` or completion of read may clash - only one has to win.